### PR TITLE
[User History] Migrate change messages to new structure & add for domain

### PR DESCRIPTION
**Pull request by mkangia**
## Summary

Data migration for three changes
1. structured change messages: https://github.com/dimagi/commcare-hq/pull/30253
2. new for_domain column: https://github.com/dimagi/commcare-hq/pull/30301
3. consistent change messages: https://github.com/dimagi/commcare-hq/pull/30253/commits/76996b5a129be4e95f5c5bedd0aba74c50088d15

For non-sass(production and staging) environments
just reset logs for all these changes.

For sass environments:

For "structured change messages" (https://github.com/dimagi/commcare-hq/pull/30253)
1. Copy all records to `LogEntry` as a backup to retain the string version of the messages for any future debugging for migration
2. Migrate the string messages to new structure messages and copy over new columns (changes, changed_via from details)
3. Old columns (details, message) to be retained for a while before deleting in a follow up PR

For "for_domain" (https://github.com/dimagi/commcare-hq/pull/30301)
1. for commcare user, always use the user's domain.
2. For web user, for updates only,
use the same domain as by_domain where they are meant to be equal (which is every place except domain membership removal, bulk upload and API updates as observed in the [PR](https://github.com/dimagi/commcare-hq/pull/30301)), so
- if API update, skip
- if bulk upload then use the record result to find the domain passed for the user in the upload, if no domain was passed, set it same as by_domain as seen for [web user bulk upload](https://github.com/dimagi/commcare-hq/blob/ae900958c257048a7e1b1e3b3db2ae9eca08c157/corehq/apps/user_importer/importer.py#L655) and [mobile user bulk upload](https://github.com/dimagi/commcare-hq/blob/ae900958c257048a7e1b1e3b3db2ae9eca08c157/corehq/apps/user_importer/importer.py#L449)
- if log is for domain removal, use the change message to extract for_domain
- else use by_domain if available

For "consistent change messages" (https://github.com/dimagi/commcare-hq/pull/30253/commits/76996b5a129be4e95f5c5bedd0aba74c50088d15)
Filter for logs which have changes for location or assigned locations and add to `change_messages` if not already present.
To be run after structured change messages has finished.

Roll out: 
run the management commands post deploy of related PRs, in order
    structured change messages
    new for_domain column
    consistent change messages
In this interval after deploy and before migration, only the User History report will get effected which is only available internally anyway.

## Feature Flag

`USER_HISTORY_REPORT`

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Added tests for the migration

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-3465

### Safety story

Apart from a real run on staging, a dry run (logging without real updates) is being done on production to be sure most messages will get successfully migrated and those that can't, can then be migrated manually.
1. Structured change messages
Dry run on prod
183 out of 192643 had issues
- most of them are where the names have changed so logs are being flagged
- and for others locations are now missing, so we are not able to confirm their names and hence flagged
all of these 183 would need to be migrated manually after a manual check of their change messages generated
2. new for_domain column
Dry run on prod
- for ALL commcare user logs domain was found
- for web users out of 5402 logs, 3654 are update log, for_domain was found for 3250
- failed bulk upload hit for bulk upload [19097, 19098, 19187, 19861, 19862, 19863, 19864] for which corresponding logs are like less than 20 and can be easily fixed manually
- 22 unexpected_flag_value (the flag set for each row in upload) hit which were exception messages and actually the same one for "You cannot remove a web user that is not a member or invited to this project. {username} is not a member or invited." for a single user. Its not clear why that was even logged as user history since there was an error but fine to ignore for now

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Comment by mkangia** [comment]
@esoergel @snopoke 

This is a rough one for review but I have tried to make it easier by adding detailed steps/description and also broken down into commits.
Overall this is mainly data migrations to populate data for production (and staging).
I have confidence now that all of this will go well on production. QA was already done on the bare commands (i added some options post, but I will do a dry run again and also run it on staging)

These are to be run post the release of the related PRs and there is no rush to run it immediately after deploy etc.
Features have been QA'ed and have passed so they are expected to go out in next week deploy.

I will be happy to setup some time to walk through it or answer queries on a call if needed.

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (snopoke) [inline corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:241]: won't this raise a `ResourceNotFound` error if the doc isn't found?

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (snopoke) [inline corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:640]: is this section still needed?

**Review by snopoke — APPROVED** [review]

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (mkangia) [reply corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:241]: Yes, this would and all exceptions generated in `migrate` are caught and logged.

Do you mean the following assertion for program is hence not needed? That sounds right.

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (mkangia) [reply corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:640]: I retained it for reference only, in case it is suggested that this should be done to validate correct migration and should be included.

It was run once before but did give a very noisy output since some messages translate differently now.

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (esoergel) [inline corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:144]: Might want to paginate this queryset and the other - this'll pull all records into memory at once

File: corehq/apps/users/management/commands/populate_for_domain_for_user_history.py
Comment (esoergel) [inline corehq/apps/users/management/commands/populate_for_domain_for_user_history.py:146]:
```suggestion
```

File: corehq/apps/users/management/commands/populate_for_domain_for_user_history.py
Comment (esoergel) [inline corehq/apps/users/management/commands/populate_for_domain_for_user_history.py:147]: What is this file?  Mobile worker info gets added only when this can't figure out the domain, but web users only get added when you can?

Comment (esoergel) [reply corehq/apps/users/management/commands/populate_for_domain_for_user_history.py:147]: Oh actually I see this block is executed for both mobile and web users.  I still don't follow what it's doing

**Review by esoergel — COMMENTED** [review]
This code would be easier to reason about if it were factored apart to a greater degree.  If you need very large if/else blocks, they can be isolated into functions so as to limit the scope of local variables they can impact (ie, each branch can be considered independently.  With mixed levels of indentation, it's difficult to tell what scope a section of code is operating in.

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (mkangia) [reply corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:170]: Just if block, yes. It would always be `True` after the [migration in the base PR](https://github.com/dimagi/commcare-hq/pull/30253/files#diff-3efade2c2ce1fb66eb147d588d776c09efd8bc0cb22db2ecdc6fc68ad07e8d79R16) is run

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (mkangia) [reply corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:253]: > Why are these fallbacks commented out?

my intention was to document that there is still a back up to simply populate the necessary information in case the object is missing. This is true for all places where we are fetching objects i.e locations, groups etc.

> it seems fine to just do this up-front rather than re-fetching the model

The re-fetch is done as a validation to ensure that the message was parsed correctly. Since we are dealing with regex and reading string messages, migrating ~200k records, it would not be possible to do any confident sanity check.

>  Since this is a one-off migration and won't stay in the codebase long

That is correct. I would be deleting these files post run on production.
It was possible to have a separate branch for just the django migration file, merge that and keep these migration files in a separate branch, setup release, run and not merge it at all.
But since they were related i kept them together in the same PR. Happy to go to the other route if that makes sense 
from a better git history.

File: corehq/apps/users/management/commands/populate_for_domain_for_user_history.py
Comment (mkangia) [reply corehq/apps/users/management/commands/populate_for_domain_for_user_history.py:15]: Sorry, That is confusing. I will change these to by_domain and `change_messages` before merging

File: corehq/apps/users/management/commands/populate_for_domain_for_user_history.py
Comment (mkangia) [reply corehq/apps/users/management/commands/populate_for_domain_for_user_history.py:146]: The migration would always log the updates but only actually update the records if `--save` is passed.
So, this is where we log into a file what the update was/would be.

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (esoergel) [reply corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py:251]: > The re-fetch is done as a validation to ensure that the message was parsed correctly. Since we are dealing with regex and reading string messages, migrating ~200k records, it would not be possible to do any confident sanity check.

That sounds reasonable

File: corehq/apps/users/management/commands/populate_for_domain_for_user_history.py
Comment (esoergel) [reply corehq/apps/users/management/commands/populate_for_domain_for_user_history.py:144]: I don't think it'll log updates for web users where `for_domain` is falsy.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/users/management/commands/populate_for_domain_for_user_history.py
Comment (mkangia) [reply corehq/apps/users/management/commands/populate_for_domain_for_user_history.py:144]: Ya, that sounds right no?
if there is no for_domain, there wont be any update.

File: corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py
Comment (mkangia) [reply corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py]: https://github.com/dimagi/commcare-hq/commit/452498d5477e6676cffd0ed407da9faf4f20ad7c

**Comment by mkangia** [comment]
@esoergel 
As discussed offline on web-team, [updated to not rely on the settings var](https://github.com/dimagi/commcare-hq/commit/60ceea3f81e6274094861ced1e349662b2eb9c2e)
Going to merge this post production deploy and then setup release to fake the migration instead.

Also updated to use iterator as suggested.

**Comment by mkangia** [comment]
Production deploy completed 42 mins ago

**Pull request merged**
